File: packages/sdk-python/examples/04_streaming.py
"""04_streaming.py — Stream answer tokens as the model emits them.

Demonstrates: AsyncClient.query.stream(), early-termination via
break, the canonical SSE event taxonomy.

Env: same as 01_quick_start.py.

Run:
    python 04_streaming.py

See also:
    docs/development/sdks/SDK_COOKBOOK_OUTLINE.md §5
"""

from __future__ import annotations

import asyncio
import os
import sys

from textral import AsyncClient

PROVIDER_KEY_REF = os.environ.get("TEXTRAL_PROVIDER_KEY_REF", "openai")
NAMESPACE = os.environ.get("TEXTRAL_NAMESPACE", "cookbook")


async def main() -> None:
    async with AsyncClient() as client:
        token_count = 0
        async for frame in client.query.stream(
            namespace=NAMESPACE,
            query="Who was Hypatia, and what happened to her?",
            embedding={
                "provider": "openai",
                "model": "text-embedding-3-large",
                "dimensions": 1536,
                "provider_key_ref": PROVIDER_KEY_REF,
            },
            inference={
                "provider": "openai",
                "model": "gpt-4o-mini",
                "provider_key_ref": PROVIDER_KEY_REF,
            },
        ):
            event = frame.get("event")
            if event == "answer_token":
                sys.stdout.write(frame.get("value", ""))
                sys.stdout.flush()
                token_count += 1
            elif event == "citation":
                # Render each citation as we receive it; alternative
                # is to wait for `final` and render the whole list.
                pass
            elif event == "final":
                print()
                print(f"--- {token_count} tokens streamed ---")
                print(f"query_event_id: {frame.get('query_event_id')}")
                print(f"citations:      {len(frame.get('citations') or [])}")


if __name__ == "__main__":
    asyncio.run(main())

File: packages/sdk/examples/04-streaming.ts
/**
 * 04-streaming.ts — SSE streaming query.
 *
 * Demonstrates: client.query.stream(...), AbortSignal cancellation,
 * TextralStreamInterrupted handling.
 *
 * Env: same as 01-quick-start.ts.
 *
 * Run:
 *     npx tsx 04-streaming.ts
 *
 * See also:
 *     SDK_COOKBOOK_OUTLINE.md §5
 */

import { TextralClient, TextralStreamInterrupted } from '@textral/sdk';

const profileName = process.env.TEXTRAL_PROFILE;
const namespace = process.env.TEXTRAL_NAMESPACE ?? 'cookbook';
const providerKeyRef = process.env.TEXTRAL_PROVIDER_KEY_REF ?? 'openai';

const client = profileName
  ? new TextralClient({ profile: profileName })
  : new TextralClient({
      baseUrl: process.env.TEXTRAL_BASE_URL ?? 'https://api.textral.alacrity.ai',
      apiKey: process.env.TEXTRAL_API_KEY ?? '',
    });

const ac = new AbortController();
// Safety timeout — cancel after 30s no matter what.
setTimeout(() => ac.abort(), 30_000).unref();

try {
  for await (const frame of client.query.stream(
    {
      namespace,
      query: 'What survived the Library of Alexandria?',
      embedding: {
        provider: 'openai',
        model: 'text-embedding-3-large',
        dimensions: 1536,
        provider_key_ref: providerKeyRef,
      },
      inference: {
        provider: 'openai',
        model: 'gpt-4o-mini',
        provider_key_ref: providerKeyRef,
      },
    },
    { signal: ac.signal },
  )) {
    switch (frame.type) {
      case 'token':
        process.stdout.write(frame.value);
        break;
      case 'citation':
        console.error(`\n  [${frame.ordinal}] ${frame.section_path}`);
        break;
      case 'audit':
        console.error(`\n  audit: ${frame.query_event_id}`);
        break;
      case 'done':
        // streamSse returns on [DONE]; this is mostly informational.
        break;
    }
  }
  console.log();
} catch (e) {
  if (e instanceof TextralStreamInterrupted) {
    console.error('stream interrupted:', e.message);
    process.exit(2);
  }
  throw e;
}
